### DISCUSSION TITLE
@on() Handlers behave differently after upgrade from WAVE 0.24.2 to 0.25.2

### DISCUSSION THREAD

**Discussion by Nayananga** [h2oai/wave#1901] [Q&A]
In wave 0.24.2

https://user-images.githubusercontent.com/25351304/227117736-9ebd55a4-ef06-404f-9f47-2187f02ade62.mov

in wave 0.25.2

https://user-images.githubusercontent.com/25351304/227117832-b1ee0a51-43ac-4925-848b-fe2a67218bb6.mov

sample code:

```
from h2o_wave import main, app, Q, ui, on, handle_on


@on()
async def button_click(q: Q):
    q.page['meta'] = ui.meta_card(box='', notification_bar=ui.notification_bar(
        text='you clicked button',
        type='success',
        position=ui.NotificationBarPosition.TOP_CENTER,
    ))

    await q.page.save()


@on("#hash_route")
async def dropdown_handle(q: Q):
    q.page['meta'] = ui.meta_card(box='', notification_bar=ui.notification_bar(
        text='you clicked hash route',
        type='success',
        position=ui.NotificationBarPosition.BOTTOM_CENTER,
    ))
    await q.page.save()


@app('/demo')
async def serve(q: Q):
    await handle_on(q)
    q.page['meta'] = ui.form_card(box='1 1 2 2', items=[
        ui.button(name='button_click', label='Button Click'),
        ui.button(name='#hash_route', label='Change Hash route'),
    ])

    await q.page.save()
```

This behavior heavily affects apps that already use `@on()` handlers, they need to be completely revamped and tested. Any thoughts?

**Comment by mturoci** [comment]
The current `handle_on` mechanism is about to be deprecated in (hopefully) next version as it has a lot of design flaws, on of which being dependant on the order of q.args/q.events. I recommend using either some custom solution or good old if else statements until then.

Note that your app should never depend on the order of q.args/q.events as it's not guaranteed. Instead, you should handle them all regardless of the order they come - handle them as batches rather than individuals.

**Reply by dulajra** [reply]
@mturoci I also agree that we should never depend on the order of q.args because that can cause unexpected behaviours in the app. But the `handle_on` was a major feature in of Wave and both of our apps (MLOps app and admin analytics app) heavily depend on it. So removing it from the framework will have a big impact on us. And if you are going to remove such a major feature from the framework then you should introduce a better alternative for the same purpose. Otherwise, it will be a big impact for all the framework users. 

> Instead, you should handle them all regardless of the order they come - handle them as batches rather than individuals.

Can you please elaborate? Actually, this should be part of the wave official docs. The docs should guide the users on best practices to handle the events within apps. 

> I recommend using either some custom solution or good old if else statements until then.

Using if else will never be an option because it will be a nightmare to manage when it comes to apps with many user actions. This will work for data visualisation apps with relatively lesser user actions. And using if else won't solve the problem of order of args. 

cc: @ShehanIshanka @senalw @VijithaEkanayake @shereezhang

**Reply by mturoci** [reply]
> And if you are going to remove such a major feature from the framework then you should introduce a better alternative for the same purpose

Sure, deprecation means using the current `handle_on` will be discouraged and replaced with a better alternative.

> Can you please elaborate? Actually, this should be part of the wave official docs. The docs should guide the users on best practices to handle the events within apps.

I am not sure what best practices you mean. My advice is common knowledge when working in event-based systems. There is nothing Wave-specific.

Btw, we have a best practice documentation - https://wave.h2o.ai/docs/routing/#organizing-code already, it just seems to be ignored by some folks.

> Using if else will never be an option because it will be a nightmare to manage when it comes to apps with many user actions.

I would like to hear more about what makes you think that `if` statements are unmanageable (note that `handle_on` is doing the very same under the hood as well). Considering the following 2 code samples.

```py
from h2o_wave import main, app, Q, ui, handle_on, on


@on('button1')
async def handle_button1(q: Q):
    pass


@on('button2')
async def handle_button2(q: Q):
    pass


@on('button3')
async def handle_button3(q: Q):
    pass


@app('/')
async def serve(q: Q):
    await handle_on(q)
    await q.page.save()
```

```py
from h2o_wave import main, app, Q, ui


async def handle_button1(q: Q):
    pass


async def handle_button2(q: Q):
    pass


async def handle_button3(q: Q):
    pass


@app('/')
async def serve(q: Q):
    if q.args.button1:
        await handle_button1(q)
    elif q.args.button2:
        await handle_button2(q)
    elif q.args.button2:
        await handle_button3(q)

    await q.page.save()
```

What exactly is better about the first approach? From my point of view, the second code will always be preferred because:

* It's clear what is going on, what will trigger, and when - super simple to debug.
* Flexibility - gives the developer the control to change the order according to their specific needs.
* Self-explanatory - one can understand the code without reading the docs explicitly.

These are the 3 main reasons why I or @lo5 always prefer the `ifelse` statements. I would only use `handle_on` if I needed some pattern matching and was too lazy to code my own.

> using if else won't solve the problem of order of args.

It does, since the developer is in charge of the event handling order unlike with `handle_on`.

**Reply by dulajra** [reply]
> Sure, deprecation means using the current handle_on will be discouraged and replaced with a better alternative.

@mturoci Great! When will that alternative be available in the SDK?

**Reply by dulajra** [reply]
> I would like to hear more about what makes you think that if statements are unmanageable

The wave [docs](https://wave.h2o.ai/docs/routing#reducing-boilerplate) answers your question very well. 

```
As your application gets larger, using the above if/elif/else conditionals can seem tedious or repetitive. If so, you can use on and handle_on to reduce the boilerplate.
```

**Reply by dulajra** [reply]
> What exactly is better about the first approach? From my point of view, the second code will always be preferred because:
> 
> - It's clear what is going on, what will trigger, and when - super simple to debug.
> - Flexibility - gives the developer the control to change the order according to their specific needs.
> - Self-explanatory - one can understand the code without reading the docs explicitly.

This is true for smaller apps or apps with less number of user interactions (events). But I would say both the mlops wave app and the admin analytics apps are bigger and complex enough to become unmanageable with a series of if-else-if statements. It will kill the manageability and the extendability of the app as well. 

cc: @VijithaEkanayake @ShehanIshanka @senalw @mwysokin

**Reply by mturoci** [reply]
All of the points above are generic and apply to any size of the application. I have still not heard a single concrete argument on why the if else statements are unmanageable. The amount of interactions doesn't matter at all, what matters is cyclomatic complexity of the code which makes it hard to maintain.

**Reply by dulajra** [reply]
@mturoci Then are you saying what is mentioned in the [docs](https://wave.h2o.ai/docs/routing#reducing-boilerplate) are also invalid?

**Reply by dulajra** [reply]
> note that handle_on is doing the very same under the hood as well

If that is the case why are you planning to remove `handle_on` from the framework and recommend using if-else for the developers? You can keep hande_on and let the developer to decide what to use for their use case.

**Reply by mturoci** [reply]
> Then are you saying what is mentioned in the [docs](https://wave.h2o.ai/docs/routing#reducing-boilerplate) are also invalid?

Why? All it says is that if it seems repetitious to you, you can use `handle_on`. It doesn't say it's the preferred way for large apps. We explicitly say what is our recommendation regarding code structure in https://wave.h2o.ai/docs/routing/#organizing-code. 

> If that is the case why are you planning to remove handle_on from the framework and recommend using if-else for the developers? 

Please read my comments thoroughly. I said we are going to replace handle_on with a newer, better version. Removal is not mentioned at all.

> You can keep hande_on and let the developer to decide what to use for their use case.

Again, read what I wrote above. This is what is happening.

**Reply by dulajra** [reply]
> The amount of interactions doesn't matter at all, what matters is cyclomatic complexity of the code which makes it hard to maintain.

That's what I meant also. The amount of interactions is a direct contribution to increase the cyclomatic complexity of the code if we are going to use if-else for each event. And there is another problem contributing to the same factor because of the design of wave. If we consider any other framework like React or Angular then we can register listeners for events. Each listener is responsible for taking care of that particular event. But in Wave, there is no way to decide the exact event which triggered that serve cycle. We need to do so many state comparisons manually to decide which event caused the serve cycle to trigger. These state checks will also contribute = to the cyclomatic complexity.

**Reply by mturoci** [reply]
> The amount of interactions is a direct contribution to increase the cyclomatic complexity of the code if we are going to use if-else for each event

This is not true. This would be only problematic if you have deep if else nesting => spaghetti code. Having a single level, flat if else statements is a no-brainer + gives you flexibility to build your own mechanism around that (you can even get rid of if else statements completely and split them into dedicated functions, bring your own event processing mechanism tailored to exactly to your needs) + other benefits mentioned above.

>  If we consider any other framework like React or Angular then we can register listeners for events.

I suggest digging deeper than just syntax level. Then you will learn Wave behaves exactly like these 2 (or JS in general) frameworks. How do you think events from React get into your defined handler? The only difference is that Wave provides finer control over triggering handlers which is better for every experienced developer.

**Comment by mtanco** [comment]
Order of `q.args` is never guaranteed, so in general this is “risky” practice to assume that the event you want will be first in line. In wave 0.25.2 there was a change that moved # values to the front of the q.args list. Because of https://github.com/h2oai/wave/issues/769, the # value is in q.args even if it has not changed. For this reason, the # `on` function is getting called instead of the button that the user is clicking on. 

The location of # in q.args can be reverted to pre-0.25.2, **but this does not change that q.args order cannot be guaranteed**. This will be applied to 0.25.3 which can be cut early week of 05/08.


Handle_on has a series of problems that developers have ran into, but are working around and using today in their apps. To improve this functionality without breaking existing apps, a new method (something like "handle_change") will be created in the coming weeks. After developers have had sufficient time to move over, the handle_on functionality will be depreciated and removed.
